File: src/0841_keys_and_rooms/keys_and_rooms.go
// author: yqq
// date: 2021-09-18 14:46:04
// descriptions:
package main
import "fmt"

func dfs(key int, rooms [][]int, visited []bool ) {
	if visited[key] {
		return;
	}

	visited[key] = true
	keys := rooms[key]
	for _ , key := range keys {
		// 深度优先搜索遍历
		dfs(key, rooms, visited);
	}
}

func canVisitAllRooms(rooms [][]int) bool {

	visited := make([]bool, len(rooms));
	dfs(0, rooms, visited);

	//检查是否都访问到了
	for i := 0; i < len(visited); i++ {
		if !visited[i] {
			return false;
		}
	}
	return true;
}


func main() {
    fmt.Println("hello world")
}

File: src/0841_keys_and_rooms/keys_and_rooms.java
import java.util.List;
import java.util.ArrayList;

class Solution {

    private void dfs(int key, List<List<Integer>>  rooms, List<Boolean> visited) {
        if (visited.get(key)) {
            return;
        }

        visited.set(key, true);
        for (int k : rooms.get(key)) {
            // 深度优先搜索遍历
            dfs(k, rooms, visited);
        }
    }


    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        List<Boolean> visited = new ArrayList<Boolean>(){{
            for(int i = 0 ; i < rooms.size(); i++){
                add(false);
            }
        }};

        dfs(0, rooms, visited);
        //检查是否都访问到了
        for (boolean flag : visited) {
            if (!flag) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        // int[][] data = new int[][]{{1},{2},{3},{}};
        List<List<Integer>> rooms = new ArrayList<>();
        rooms.add(new ArrayList<Integer>(){ {add(1);} });
        rooms.add(new ArrayList<Integer>(){ {add(2);} });
        rooms.add(new ArrayList<Integer>(){ {add(3);} });
        rooms.add(new ArrayList<Integer>(){ {} });
        // System.out.println(rooms.size() );
        System.out.println( solution.canVisitAllRooms(rooms) );
    }
}